sycl/doc/extensions/proposed/sycl_ext_oneapi_graph_fusion.asciidoc

```diff
@@ -162,33 +202,53 @@ this buffer will inherit this property (unless overridden).
 `sycl::aligned_alloc_device()`, `sycl::malloc_shared()`, or
 `sycl::aligned_alloc_shared()` to apply the property to an USM pointer.
 
-```c++ 
-sycl::ext::oneapi::experimental::property::promote_local 
+```c++
+namespace sycl::ext::oneapi::experimental::property{
+
+  template<sycl::memory_scope Scope>
+  struct access_scope {};
+
+  inline constexpr auto access_scope_local = 
+                                    access_scope<memory_scope_work_group>;
+
+  inline constexpr auto access_scope_private = 
+                                    access_scope<memory_scope_work_item>;
+
+} // namespace sycl::ext::oneapi::experimental::property
 ```
 
-This property is an assertion by the application that each element in the buffer
-or allocated device memory is accessed by no more than one work-group in the
-kernel submitted by this command-group (in case the property is specified on an
-accessor) or in any kernel in the graph (in case the property is specified on a
-buffer or an USM pointer). Implementations may treat this as a hint to promote
-the elements of the buffer or allocated device memory to local memory (see below
-section on local and private internalization).
-
-The application also asserts that the updates made to the buffer or allocated
-device memory by the kernel submitted by this command-group (in case the
-property is specified on an accessor) or in any kernel in the graph (in case the
-property is specified on a buffer or an USM pointer) may not be available for
-use after the fused kernel completes execution. Implementations may treat this
-as a hint to not initialize or write back the final result to global memory.
-
-If different properties (local or private) are applied to accessors to the same
-buffer, the resolution rules specified below apply. The property is not
-prescriptive, implementations are free to not perform internalization and it is
-no error if they do not perform internalization. Implementations can provide a
-diagnostic message in case internalization was not performed through an
-implementation-specified mechanism, but are not required to do so.
-
-===== Private internalization property
+Specializations of the `access_scope` property template can be used to express
+the access pattern of kernels to a buffer or USM allocation.
+
+The specializations of the property are an assertion by the application that
+each element in the buffer or allocated device memory is at most accessed in
+the given memory scope in the kernel submitted by this command-group (in case
+the property is specified on an accessor) or in any kernel in the graph (in case
+the property is specified on a buffer or an USM pointer).
```

Comment:
The second half of this sentence is surprising to me.  In a real application, a developer might not have insight into or control over how a USM allocation is used by other kernels.

Wouldn't it be safer to require this property on all usages of USM within the graph?

Reply:
The issue with USM is that throughout most of the application, these are just plain C++ pointers, so it's hard to attach things to them at their usage point (there's no accessors created for USM and no `require`). 

With the standard USM API, the allocation is more or less the only place to attach a property list to a USM pointer. 

In an earlier version, we used the USM allocation nodes of the graph to narrow the scope of the USM pointer and therefore also internalization properties to just the graph, but the allocation nodes did not make it into the current version of the graph proposal. 

A potential alternative to allow more fine-grained would be to introduce properties to mark USM pointers that take the USM pointer itself: 
```c++
template<sycl::memory_scope Scope>
struct usm_access_scope {
  void* ptr;
};
```

These properties could then be passed to `command_graph::finalize()` to allow limit internalization semantics to just that one graph:
```c++
auto* usm_ptr = malloc_shared(...);
...
graph.finalize(require_fusion{}, usm_access_scope<memory_scope::work_item>{usm_ptr});
```

We could introduce this in addition to the `access_scope` property and either only allow this for USM pointers, or allow both alternatives. WDYT?

Reply:
Oh, I see.  Thank you for the explanation.  Somehow, I missed that this property was attached to the malloc.

I like your idea of being able to pass a pointer into the `usm_access_scope` property, but I'm not sure I follow why it should be part of the `finalize` call.  It sounds like what we really want to be able to do is mark individual usages, so that we can make different fusion decisions for different kernels using the same pointer(s).

What if you made it so that `access_scope` was a valid property for an [sycl_ext_oneapi_annotated_ptr](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc)?  That way, a developer could attach access scope information to a pointer and carry that information as long as necessary.  If the access scope applies to the whole graph, the annotation can be attached to the pointer once, and that same annotated pointer can be captured by all the kernels.  If the access scope applies to a single kernel, an annotated pointer can be created locally at command-group scope.

Would that make sense?

One limitation of this approach that I can see is that you'd only be able to use the information if a USM pointer was one of the kernel arguments (i.e. it couldn't be captured indirectly or via a struct).  But I suspect this may be an issue with the current proposal already, since there's no association between pointers and the kernels/graphs that use them.  Do you need to add a restriction somewhere talking about this?

Reply:
> For symmetry and because implementations might require all three properties to successfully perform internalization, `no_init` would also need to be allowed for `annotated_ptr`. I would need to check the implementation, but I assume that the `no_init` property should also fit into the compile property framework. WDYT?

I hadn't thought about this.  One problem I see here is that `no_init` for buffers/accessors is scoped to a command-group scope, so the property implicitly means "this memory is not initialized prior to this command-group".  It's not clear to me what it would mean if you attached it to a pointer at arbitrary scope.  If we attached `no_init` to an `annotated_ptr` before any kernels, the interpretation seems like it would have to be "this memory is never initialized" or "this memory is not initialized before its first use".

Do you think the second interpretation is implementable?  I can sort of see how you might be able to track this in the runtime.

> I'm not entirely sure that I get what you mean here. As fusion happens at runtime, it is usually possible to determine the arguments captured/used by a kernel through the SYCL runtime, so implementations of fusion have access to that information, unless it's nested inside structs or similar

Right, we're on the same page.  I'm suggesting that you add text to the effect of:

"The properties of an `annotated_ptr` should be inspected by an implementation when the `annotated_ptr` is captured by a kernel lambda or otherwise passed as an argument to a kernel function. Implementations are not required to track fusion-related information for other USM pointers that may be used by a kernel, such as those stored inside of structs or other data structures."

I think it should be a weak "not required" because some implementations (like DPC++) might unpack some structs in some situations.  But I think it's important to note that if a developer wants the information to be usable, they should ensure that the pointer is captured as an argument.

---

One other possible restriction came to mind while I was writing the above.  What is supposed to happen in a case like this:
```c++
int* some_regular_ptr = ...;
annotated_ptr<int, ...> some_annotated_ptr = ...; // with appropriate internalization properties

q.parallel_for(..., [=](auto& i) {
  foo(some_regular_ptr);
  bar(some_annotated_ptr);
});
```

If `some_annotated_ptr` is internalized successfully, `some_regular_ptr` still won't be.  (If you think the compiler is smart enough to detect this case, assume `some_regular_ptr` is in a struct and `some_annotated_ptr` isn't).  Am I right in thinking that any code which expects these two pointers to point to the same memory will break?

If so, I think you should add another restriction saying that this is UB.

Reply:
> > For symmetry and because implementations might require all three properties to successfully perform internalization, `no_init` would also need to be allowed for `annotated_ptr`. I would need to check the implementation, but I assume that the `no_init` property should also fit into the compile property framework. WDYT?
> 
> I hadn't thought about this. One problem I see here is that `no_init` for buffers/accessors is scoped to a command-group scope, so the property implicitly means "this memory is not initialized prior to this command-group". It's not clear to me what it would mean if you attached it to a pointer at arbitrary scope. If we attached `no_init` to an `annotated_ptr` before any kernels, the interpretation seems like it would have to be "this memory is never initialized" or "this memory is not initialized before its first use".
> 
> Do you think the second interpretation is implementable? I can sort of see how you might be able to track this in the runtime.

What we really want to express with the `no_init` property in the context of fusion is that before any value of that buffer is read, it is written by the same kernel or another kernel that will come before it in the fused kernel. 

To that end, it is not really important if the memory is initialized on the host before use or whether the initialized memory is copied to the device. What we want to achieve with this property is that the fusion compiler can assume the first memory operation to that buffer (if any) will be a write. Otherwise, the compiler would need to identify the first loads from the buffer on any path of execution (which requires non-trivial analysis) and not internalize those loads, in order to initialize the internalized memory (register or local memory) with the value from global memory.

In my interpretation of the spec, this would be covered by the semantics of the `no_init` property, but I see your point about potential further reaching semantic implications for USM pointers. 

Alternatively, we could introduce another more fusion/graph-specific property (similar to `fusion_internal_memory`) to make it clearer. Which solution would you prefer? 

> 
> > I'm not entirely sure that I get what you mean here. As fusion happens at runtime, it is usually possible to determine the arguments captured/used by a kernel through the SYCL runtime, so implementations of fusion have access to that information, unless it's nested inside structs or similar
> 
> Right, we're on the same page. I'm suggesting that you add text to the effect of:
> 
> "The properties of an `annotated_ptr` should be inspected by an implementation when the `annotated_ptr` is captured by a kernel lambda or otherwise passed as an argument to a kernel function. Implementations are not required to track fusion-related information for other USM pointers that may be used by a kernel, such as those stored inside of structs or other data structures."
> 
> I think it should be a weak "not required" because some implementations (like DPC++) might unpack some structs in some situations. But I think it's important to note that if a developer wants the information to be usable, they should ensure that the pointer is captured as an argument.

Ok, I'll add that note to the proposal. 

I think the use that we want to restrict here is also forbidden by the `annotated_ptr` extension: 
> It is illegal to apply annotated_ptr to members of kernel arguments.

But I would still add the note to make it more self-contained and in case the `annotated_ptr` proposal changes.

> 
> One other possible restriction came to mind while I was writing the above. What is supposed to happen in a case like this:
> 
> ```c++
> int* some_regular_ptr = ...;
> annotated_ptr<int, ...> some_annotated_ptr = ...; // with appropriate internalization properties
> 
> q.parallel_for(..., [=](auto& i) {
>   foo(some_regular_ptr);
>   bar(some_annotated_ptr);
> });
> ```
> 
> If `some_annotated_ptr` is internalized successfully, `some_regular_ptr` still won't be. (If you think the compiler is smart enough to detect this case, assume `some_regular_ptr` is in a struct and `some_annotated_ptr` isn't). Am I right in thinking that any code which expects these two pointers to point to the same memory will break?
> 
> If so, I think you should add another restriction saying that this is UB.

Internalization of two different pointers is independent from each other, so one pointer can be internalized, while the other remains untouched, that is no problem. 

In case `some_regular_ptr` and `some_annotated_ptr` point to the same memory, no internalization would happen. As fusion happens at runtime, it's relatively easy to detect if the two pointers are identical. In that case, the same resolution rules as for two accessors to the same buffer (second table in 7.6.1) could apply: One captured pointer has a work-group scope attached, the other not, so the first line of the table would apply and no internalization would happen. I would extend the text before that table to also talk about USM pointers. 

The only case for which we should probably add a restriction saying it's UB is if `some_annotated_ptr` is `some_regular_ptr` plus some offset. As USM pointers do not carry their allocation size, it's impossible to determine if two pointers partially overlap, and the fusion compiler should be able to assume they do not, unless they are identical.

Reply:
> Alternatively, we could introduce another more fusion/graph-specific property (similar to `fusion_internal_memory`) to make it clearer. Which solution would you prefer?

Thanks for the explanation.  I prefer using `no_init`, here -- I just want to make sure we'd thought through the corner cases, and document anything that might be important.

> > It is illegal to apply annotated_ptr to members of kernel arguments.
> 
> But I would still add the note to make it more self-contained and in case the `annotated_ptr` proposal changes.

I think that's a good idea.  I'd missed/forgotten this note in the `annotated_ptr` proposal, so highlighting it here is helpful.
 
> In case `some_regular_ptr` and `some_annotated_ptr` point to the same memory, no internalization would happen. As fusion happens at runtime, it's relatively easy to detect if the two pointers are identical. In that case, the same resolution rules as for two accessors to the same buffer (second table in 7.6.1) could apply: One captured pointer has a work-group scope attached, the other not, so the first line of the table would apply and no internalization would happen. I would extend the text before that table to also talk about USM pointers.

But this is only true if both `some_regular_ptr` and `some_annotated_ptr` are captured as arguments, isn't it?  If they point to the same memory but `some_regular_ptr` is accessed indirectly (e.g. via a struct member, a pointer to pointer, some kernel-dependent logic) then the runtime won't be able to compare the pointers.

As a somewhat realistic use-case: imagine that there's some big allocation being used as the backing store for a data structure, and a data structure layered on top (e.g. a linked list or something).  A developer can declare the big allocation as having some fusion-related properties, but those properties won't carry through to the data structure.  The runtime can't apply the accessor resolution rules, because it can't see the pointers in the data structure.  The runtime in this case might conclude that the memory _can_ be internalized, but any changes to the memory would break the data structure.

Does that make sense?  I feel like I might not be explaining this well.

> The only case for which we should probably add a restriction saying it's UB is if `some_annotated_ptr` is `some_regular_ptr` plus some offset. As USM pointers do not carry their allocation size, it's impossible to determine if two pointers partially overlap, and the fusion compiler should be able to assume they do not, unless they are identical.

I agree it would be a good idea to add a restriction about this too.

Reply:
> But this is only true if both `some_regular_ptr` and `some_annotated_ptr` are captured as arguments, isn't it? If they point to the same memory but `some_regular_ptr` is accessed indirectly (e.g. via a struct member, a pointer to pointer, some kernel-dependent logic) then the runtime won't be able to compare the pointers.
> 
> As a somewhat realistic use-case: imagine that there's some big allocation being used as the backing store for a data structure, and a data structure layered on top (e.g. a linked list or something). A developer can declare the big allocation as having some fusion-related properties, but those properties won't carry through to the data structure. The runtime can't apply the accessor resolution rules, because it can't see the pointers in the data structure. The runtime in this case might conclude that the memory _can_ be internalized, but any changes to the memory would break the data structure.
> 
> Does that make sense? I feel like I might not be explaining this well.
> 

That's a good point, I hadn't considered that case. 

As the absence of such an indirect use of the pointer would definitely be non-trivial for the compiler to prove, the most realistic option seems to be to have the user promise that no such use takes place. This would be possible by extending the semantics of one of the properties (I'm leaning towards `fusion_internal_memory`) or by introducing a new property, that essentially lets the user state that no such indirect use will take place. 

WDYT?

Reply:
> As the absence of such an indirect use of the pointer would definitely be non-trivial for the compiler to prove, the most realistic option seems to be to have the user promise that no such use takes place.

I agree, this makes sense to me.

> This would be possible by extending the semantics of one of the properties (I'm leaning towards fusion_internal_memory) or by introducing a new property, that essentially lets the user state that no such indirect use will take place.

I might be wrong about this, but I think you need to extend the semantics of _all_ the properties.  It seems like a developer could get themselves into similar trouble by accessing memory through two pointers with incompatible `access_scope`s (and/or raw pointers with no `access_scope`).  Maybe the simplest thing to do is to follow `atomic_ref`, and say something like:

"If an `annotated_ptr` is created with any of these properties, the underlying memory must only be accessed via an `annotated_ptr` with a compatible set of properties. During the lifetime of an `annotated_ptr` with any of these properties, direct access to the underlying memory via a non-annotated pointer results in undefined behavior."

Or is that too strong?

Reply:
Looks great, thank you for making these changes!